src/value_generators/include/value_generators/value_generators.hpp

```diff
@@ -69,6 +77,262 @@ class DocumentGenerator {
  */
 std::unique_ptr<DocumentGenerator> makeDoc(YAML::Node, genny::DefaultRandom&);
 
+
+/**
+ * A variant type for representing different kinds of values that can be generated by an Expression
+ * instance.
+ *
+ * The bsoncxx::types::b_document and bsoncxx::types::b_array cases that are part of
+ * bsoncxx::types::value do not allow for BSON documents or arrays to own their underlying buffer
+ * and can only be constructed from their respective view types.
+ *
+ * Note that we explicitly subclass from std::variant rather than alias it in order to define a
+ * custom std::ostream::operator<< for it.
+ */
+class Value : public std::variant<bool,
+                                  int32_t,
+                                  int64_t,
+                                  double,
+                                  std::string,
+                                  bsoncxx::types::b_null,
+                                  bsoncxx::document::view_or_value,
+                                  bsoncxx::array::view_or_value> {
+public:
+    int64_t getInt64(std::string_view name);
+
+private:
+    friend std::ostream& operator<<(std::ostream& out, const Value& value);
+};
+
+
+class Expression;
+using UniqueExpression = std::unique_ptr<Expression>;
+
+
+/**
+ * Base class for generating values.
+ *
+ * Its API is modeled off the Expression class for representing aggregation expressions in the
+ * MongoDB Server.
+ */
+class Expression {
+public:
+    using Parser = std::function<UniqueExpression(YAML::Node)>;
+
+    /**
+     * Parse `node` into one of the Expression types registered in the `parserMap`.
+     *
+     * @throws InvalidConfigurationException if `node` isn't a mapping type with a '^'-prefixed key.
+     */
+    static UniqueExpression parseExpression(YAML::Node node);
+
+    /**
+     * Parse `node` into either:
+     *   - one of the Expression types registered in the `parserMap`, or
+     *   - a DocumentExpression instance containing further nested Expression types.
+     *
+     * @throws InvalidConfigurationException if `node` isn't a mapping type.
+     */
+    static UniqueExpression parseObject(YAML::Node node);
+
+    /**
+     * Parse `node` into one of the following:
+     *   - one of the Expression types registered in the `parserMap`,
+     *   - a DocumentExpression instance containing further nested Expression types,
+     *   - an ArrayExpression instance containing a list Expression types,
+     *   - a ConstantExpression instance representing a scalar value.
+     *
+     * @throws InvalidConfigurationException if `node` is malformed.
+     */
+    static UniqueExpression parseOperand(YAML::Node node);
```

Comment:
I would say that `Expression::parseOperand()` is the most permissive in terms of what it parses. The choice of which function should be used depends on the context in the YAML configuration. For example, using `ArrayExpression::parse()` would throw an exception at `parse()`-time if it wasn't going to generate an array value rather than throwing an exception at `evaluate()`-time when calling `.getArray()` on the result.

I'm open to suggestions on how to make the choice more clear to future actor authors.